File: sample/swift/10_plasma_effect.swift
import Foundation


// 10: Sample that outputs a plasma effect as a GIF.

func run_10_plasma_effect() {
    var w: Int64 = __pytra_int(Int64(320))
    var h: Int64 = __pytra_int(Int64(240))
    var frames_n: Int64 = __pytra_int(Int64(216))
    var out_path: String = __pytra_str("sample/out/10_plasma_effect.gif")
    var start: Double = __pytra_float(__pytra_perf_counter())
    var frames: [Any] = __pytra_as_list([])
    let __step_0 = __pytra_int(Int64(1))
    var t = __pytra_int(Int64(0))
    while ((__step_0 >= 0 && t < __pytra_int(frames_n)) || (__step_0 < 0 && t > __pytra_int(frames_n))) {
        var frame: [Any] = __pytra_as_list(__pytra_bytearray((__pytra_int(w) * __pytra_int(h))))
        let __step_1 = __pytra_int(Int64(1))
        var y = __pytra_int(Int64(0))
        while ((__step_1 >= 0 && y < __pytra_int(h)) || (__step_1 < 0 && y > __pytra_int(h))) {
            var row_base: Int64 = __pytra_int((__pytra_int(y) * __pytra_int(w)))
            let __step_2 = __pytra_int(Int64(1))
            var x = __pytra_int(Int64(0))
            while ((__step_2 >= 0 && x < __pytra_int(w)) || (__step_2 < 0 && x > __pytra_int(w))) {
                var dx: Int64 = __pytra_int((__pytra_int(x) - __pytra_int(Int64(160))))
                var dy: Int64 = __pytra_int((__pytra_int(y) - __pytra_int(Int64(120))))
                var v: Any = (((sin(__pytra_float((__pytra_float((__pytra_float(x) + __pytra_float((__pytra_float(t) * __pytra_float(Double(2.0)))))) * __pytra_float(Double(0.045))))) + sin(__pytra_float((__pytra_float((__pytra_float(y) - __pytra_float((__pytra_float(t) * __pytra_float(Double(1.2)))))) * __pytra_float(Double(0.05)))))) + sin(__pytra_float((__pytra_float((__pytra_float((__pytra_int(x) + __pytra_int(y))) + __pytra_float((__pytra_float(t) * __pytra_float(Double(1.7)))))) * __pytra_float(Double(0.03)))))) + sin(__pytra_float(((sqrt(__pytra_float((__pytra_int((__pytra_int(dx) * __pytra_int(dx))) + __pytra_int((__pytra_int(dy) * __pytra_int(dy)))))) * Double(0.07)) - (__pytra_float(t) * __pytra_float(Double(0.18)))))))
                var c: Int64 = __pytra_int(__pytra_int(((v + Double(4.0)) * (__pytra_float(Double(255.0)) / __pytra_float(Double(8.0))))))
                if (__pytra_int(c) < __pytra_int(Int64(0))) {
                    c = __pytra_int(Int64(0))
                }
                if (__pytra_int(c) > __pytra_int(Int64(255))) {
                    c = __pytra_int(Int64(255))
                }
                __pytra_setIndex(frame, (__pytra_int(row_base) + __pytra_int(x)), c)
                x += __step_2
            }
            y += __step_1
        }
        frames = __pytra_as_list(frames); frames.append(__pytra_bytes(frame))
        t += __step_0
    }
    __pytra_noop(out_path, w, h, frames, [])
    var elapsed: Double = __pytra_float((__pytra_float(__pytra_perf_counter()) - __pytra_float(start)))
    __pytra_print("output:", out_path)
    __pytra_print("frames:", frames_n)
    __pytra_print("elapsed_sec:", elapsed)
}

@main
struct Main {
    static func main() {
        run_10_plasma_effect()
    }
}

File: sample/go/10_plasma_effect.go
package main

import (
    "math"
)


// 10: Sample that outputs a plasma effect as a GIF.

func run_10_plasma_effect() {
    var w int64 = int64(320)
    var h int64 = int64(240)
    var frames_n int64 = int64(216)
    var out_path string = __pytra_str("sample/out/10_plasma_effect.gif")
    var start float64 = __pytra_perf_counter()
    var frames []any = __pytra_as_list([]any{})
    for t := int64(0); t < frames_n; t += 1 {
        var frame []any = __pytra_as_list(__pytra_bytearray((w * h)))
        for y := int64(0); y < h; y += 1 {
            var row_base int64 = (y * w)
            for x := int64(0); x < w; x += 1 {
                var dx int64 = (x - int64(160))
                var dy int64 = (y - int64(120))
                var v float64 = (((math.Sin(((float64(x) + (float64(t) * float64(2.0))) * float64(0.045))) + math.Sin(((float64(y) - (float64(t) * float64(1.2))) * float64(0.05)))) + math.Sin(((float64((x + y)) + (float64(t) * float64(1.7))) * float64(0.03)))) + math.Sin(__pytra_float(((math.Sqrt(float64(((dx * dx) + (dy * dy)))) * float64(0.07)) - (float64(t) * float64(0.18))))))
                var c int64 = __pytra_int(((v + float64(4.0)) * (float64(255.0) / float64(8.0))))
                if (c < int64(0)) {
                    c = int64(0)
                }
                if (c > int64(255)) {
                    c = int64(255)
                }
                __pytra_set_index(frame, (row_base + x), c)
            }
        }
        frames = append(frames, __pytra_bytes(frame))
    }
    __pytra_save_gif(out_path, w, h, frames, __pytra_grayscale_palette(), int64(3), int64(0))
    var elapsed float64 = (__pytra_perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", frames_n)
    __pytra_print("elapsed_sec:", elapsed)
}

func main() {
    run_10_plasma_effect()
}
